### DISCUSSION TITLE
Add support for some JSDocs declarations such as @throws and @deprecated

### DISCUSSION THREAD

**Discussion by amoreaulemay** [biomejs/biome#2118] [Ideas]
I think it could be useful to add support for some JSDocs rules that don't have a direct TypeScript equivalent, such as `@throws` and `@deprecated`.

# `@throws`

This could be two folds:

1. Force error handling for consumers of functions that may throw. It could even force exhaustive error handling too. It could be a warning by default maybe.
2. Force declaring JSDocs with `@throws` when there's an explicit `throw` and/or it rethrows an error.

## Example

```javascript
/**
 * @throws {MyCustomError1} Thrown when a == 1 // <- Here biome forces the `@throws` JSDoc declaration; and must be exhaustive
 * @throws {MyCustomError2} Thrown when a != 1
 */
function myFunction(a) {
  if (a === 1) {
    throw new MyCustomError1()
  }

  throw new MyCustomError2()
}

function myConsumer() {
  try { // <- Here biome forces handling
    myFunction(1)
  } catch (error) { // <- Exhaustive error handling
    if (error instanceof MyCustomError1) {
      // ...handling
    } else if (error instanceof MyCustomError2) {
      // ...handling
    } else {
      // Leaving a default for other error types
    }
  }
}
```

# `@deprecated`

Self explanory but consumers of functions with a JSDoc `@deprecated` declaration should trigger a biome warning/error.

## Example

```javascript
/**
 * @deprecated - Don't use this!
 */
function dontUseThis() {
  // ...
}

function myConsumer() {
  dontUseThis() // <- Here biome triggers a warning
}
```

**Comment by arendjr** [comment]
I think both would be useful to have, but I'm not sure we currently have the infrastructure to do something like this. I think once we start implementing type inference, we may have the ability to store such annotations as well. Maybe @Conaclos has some insight here :)

**Reply by amoreaulemay** [reply]
@arendjr any update if the current architecture of Biome could support features like this?

**Reply by arendjr** [reply]
I'd say we're _almost_ there :) We track JSDoc comments now for anything that's exported, so if you were to write a rule today for this, you could actually get it to work, but you'd need to look up the JSDoc comment yourself from the CST when referencing a function within the same module (since those might not be exported), while you could use the JSDoc stored in our module graph for any functions that you import.

So if someone were very eager, they can do it today, but realistically I'd advise waiting a little bit longer. I expect this discrepancy between local and imported symbols to disappear as a consequence of the type inference work I'm doing now, so it'll actually get easier still. Feel free to remind me in a month or 2-3, then we can probably create tasks for these.

**Reply by amoreaulemay** [reply]
Thanks for the update @arendjr ! :)

**Reply by neemski** [reply]
Hey @arendjr - just curious any update here? Would be very useful to have biome warn on `@deprecated` code usage

**Reply by arendjr** [reply]
You're in luck! I added a rule for @deprecated just yesterday, though it's focused on imports only: https://github.com/biomejs/biome/pull/7520

Let me know if that suits you!

**Reply by amoreaulemay** [reply]
Awesome @arendjr ! That's a big improvement

**Reply by neemski** [reply]
Great timing! 😄 Thanks so much @arendjr this is huge! 🙏

**Reply by RedEagle-dh** [reply]
Sounds great! Excited to try it out! 👍🏽

**Comment by redbmk** [comment]
`@todo` would be another cool one to support, so we could flag it as a warning and make TODOs easier to find

**Comment by RedEagle-dh** [comment]
I would love to see the warning if using a method or function decelerated with `@deprecated`. This would help to see very early which function is marked as deprecated and will be removed in the future!

**Comment by AstroGD** [comment]
+1 Would love to see this, especially for deprecated code

**Comment by wottpal** [comment]
Same here! Current only solution is to fall back to eslint :/
